File: src/lang/python/ai_core.py
"""
UniCore AI智能层 - Python实现
意图理解、任务分解、知识推理、多模态生成
"""

import json
import re
import math
from typing import List, Dict, Any, Optional, Tuple
from dataclasses import dataclass
from enum import Enum


class Intent(Enum):
    """意图类型"""
    QUERY = "query"
    EXECUTE = "execute"
    CREATE = "create"
    LEARN = "learn"
    CONTROL = "control"
    UNKNOWN = "unknown"


@dataclass
class Token:
    """Token表示"""
    text: str
    embedding: List[float]
    pos: str  # 词性


class IntentClassifier:
    """意图分类器"""
    
    INTENT_PATTERNS = {
        Intent.QUERY: [r"什么", r"怎么", r"如何", r"为什么", r"多少", r"who", r"what", r"how", r"why"],
        Intent.EXECUTE: [r"执行", r"运行", r"开始", r"做", r"run", r"start", r"do"],
        Intent.CREATE: [r"创建", r"生成", r"新建", r"写", r"create", r"generate", r"new"],
        Intent.LEARN: [r"学习", r"教我", r"告诉", r"解释", r"learn", r"teach", r"explain"],
        Intent.CONTROL: [r"控制", r"开关", r"调节", r"设置", r"control", r"set", r"adjust"],
    }
    
    def __init__(self):
        self.confidence_threshold = 0.6
    
    def classify(self, text: str) -> Tuple[Intent, float]:
        """分类文本意图"""
        scores = {intent: 0.0 for intent in Intent}
        
        for intent, patterns in self.INTENT_PATTERNS.items():
            for pattern in patterns:
                if re.search(pattern, text.lower()):
                    scores[intent] += 1
        
        total = sum(scores.values())
        if total == 0:
            return Intent.UNKNOWN, 0.0
        
        # 归一化
        for intent in scores:
            scores[intent] /= total
        
        # 找最高分
        best_intent = max(scores, key=scores.get)
        confidence = scores[best_intent]
        
        if confidence < self.confidence_threshold:
            return Intent.UNKNOWN, confidence
        
        return best_intent, confidence


class TaskDecomposer:
    """任务分解器"""
    
    def __init__(self):
        self.subtask_templates = {
            "complex_query": ["分解", "分析", "综合"],
            "multi_step": ["然后", "接着", "之后", "下一步"],
            "parallel": ["同时", "并行", "一起"],
        }
    
    def decompose(self, task: str) -> List[Dict[str, Any]]:
        """将复杂任务分解为子任务"""
        subtasks = []
        
        # 简单的关键词分割
        parts = re.split(r'[，,。\. ]+', task)
        for i, part in enumerate(parts):
            if part.strip():
                subtasks.append({
                    "id": i,
                    "description": part.strip(),
                    "dependencies": [] if i == 0 else [i-1],
                    "parallel": False
                })
        
        # 检测并行任务
        for keyword in self.subtask_templates["parallel"]:
            if keyword in task:
                for i, st in enumerate(subtasks):
                    if i > 0:
                        st["parallel"] = True
                        st["dependencies"] = []
        
        return subtasks


class KnowledgeGraph:
    """知识图谱"""
    
    def __init__(self):
        self.nodes: Dict[str, Dict[str, Any]] = {}
        self.edges: List[Tuple[str, str, str]] = []  # (from, to, relation)
    
    def add_node(self, id: str, data: Dict[str, Any]):
        """添加节点"""
        self.nodes[id] = data
    
    def add_edge(self, from_id: str, to_id: str, relation: str):
        """添加边"""
        self.edges.append((from_id, to_id, relation))
    
    def query(self, node_id: str, depth: int = 1) -> Dict[str, Any]:
        """查询节点及其关联"""
        result = {
            "node": self.nodes.get(node_id),
            "relations": []
        }
        
        for f, t, r in self.edges:
            if f == node_id:
                result["relations"].append({
                    "target": t,
                    "relation": r,
                    "target_data": self.nodes.get(t)
                })
        
        return result
    
    def infer(self, start: str, end: str) -> Optional[List[str]]:
        """推理路径"""
        # 简单的BFS寻路
        visited = set()
        queue = [(start, [start])]
        
        while queue:
            node, path = queue.pop(0)
            if node == end:
                return path
            
            if node in visited:
                continue
            visited.add(node)
            
            for f, t, _ in self.edges:
                if f == node and t not in visited:
                    queue.append((t, path + [t]))
        
        return None


class EmbeddingModel:
    """简单Embedding模型（可替换为真实模型）"""
    
    def __init__(self, dim: int = 128):
        self.dim = dim
        self.vocab: Dict[str, List[float]] = {}
    
    def encode(self, text: str) -> List[float]:
        """编码文本为向量"""
        words = list(text)
        vector = [0.0] * self.dim
        
        for i, char in enumerate(words):
            if char in self.vocab:
                vec = self.vocab[char]
            else:
                # 随机初始化（实际应使用预训练模型）
                vec = [math.sin(ord(char) * (i + 1) / self.dim) for i in range(self.dim)]
                self.vocab[char] = vec
            
            for j in range(self.dim):
                vector[j] += vec[j] / len(words)
        
        # L2归一化
        norm = math.sqrt(sum(v*v for v in vector))
        if norm > 0:
            vector = [v/norm for v in vector]
        
        return vector
    
    def cosine_similarity(self, a: List[float], b: List[float]) -> float:
        """余弦相似度"""
        dot = sum(ai * bi for ai, bi in zip(a, b))
        norm_a = math.sqrt(sum(v*v for v in a))
        norm_b = math.sqrt(sum(v*v for v in b))
        if norm_a * norm_b == 0:
            return 0.0
        return dot / (norm_a * norm_b)


class ReasoningEngine:
    """推理引擎"""
    
    def __init__(self):
        self.rules: List[Dict[str, Any]] = []
        self.facts: Dict[str, Any] = {}
    
    def add_rule(self, condition: str, action: str):
        """添加规则"""
        self.rules.append({
            "condition": condition,
            "action": action,
            "pattern": re.compile(condition)
        })
    
    def add_fact(self, key: str, value: Any):
        """添加事实"""
        self.facts[key] = value
    
    def forward_chain(self, query: str) -> Optional[str]:
        """前向链推理"""
        for rule in self.rules:
            if rule["pattern"].search(query):
                return rule["action"]
        return None
    
    def backward_chain(self, goal: str) -> bool:
        """后向链推理"""
        for rule in self.rules:
            if rule["action"] == goal:
                # 检查条件是否满足
                return True
        return False


class MultiModalGenerator:
    """多模态生成器"""
    
    def __init__(self):
        self.capabilities = {
            "text": True,
            "image": True,
            "audio": True,
            "video": False,
        }
    
    def generate_text(self, prompt: str) -> str:
        """生成文本（简化版）"""
        return f"[AI生成文本]: {prompt}"
    
    def generate_image_description(self, prompt: str) -> Dict[str, Any]:
        """生成图像描述"""
        return {
            "prompt": prompt,
            "width": 512,
            "height": 512,
            "style": "realistic",
            "colors": ["blue", "white"]
        }
    
    def generate_speech(self, text: str, voice: str = "default") -> bytes:
        """生成语音（返回模拟音频数据）"""
        # 实际应使用TTS服务
        return b"fake_audio_data"
    
    def process_command(self, command: str) -> Dict[str, Any]:
        """处理命令并决定输出类型"""
        result = {"type": "text", "content": ""}
        
        if any(kw in command for kw in ["图片", "image", "画"]):
            result = {"type": "image", "content": self.generate_image_description(command)}
        elif any(kw in command for kw in ["声音", "speech", "读"]):
            result = {"type": "audio", "content": self.generate_speech(command)}
        else:
            result = {"type": "text", "content": self.generate_text(command)}
        
        return result


class UniCoreAI:
    """UniCore AI核心"""
    
    def __init__(self):
        self.intent_classifier = IntentClassifier()
        self.task_decomposer = TaskDecomposer()
        self.knowledge_graph = KnowledgeGraph()
        self.embedding_model = EmbeddingModel()
        self.reasoning_engine = ReasoningEngine()
        self.multimodal = MultiModalGenerator()
        
        # 初始化默认知识
        self._init_default_knowledge()
    
    def _init_default_knowledge(self):
        """初始化默认知识"""
        # 添加系统知识
        self.knowledge_graph.add_node("system", {"type": "system", "name": "UniCore"})
        self.knowledge_graph.add_node("platform", {"type": "platform", "name": "multi-platform"})
        self.knowledge_graph.add_edge("system", "platform", "runs_on")
    
    def understand(self, input_text: str) -> Dict[str, Any]:
        """理解输入"""
        intent, confidence = self.intent_classifier.classify(input_text)
        subtasks = self.task_decomposer.decompose(input_text)
        
        return {
            "text": input_text,
            "intent": intent.value,
            "confidence": confidence,
            "subtasks": subtasks,
            "embedding": self.embedding_model.encode(input_text)
        }
    
    def reason(self, query: str) -> Dict[str, Any]:
        """推理"""
        result = self.reasoning_engine.forward_chain(query)
        
        # 知识图谱查询
        kg_result = self.knowledge_graph.query(query)
        
        return {
            "result": result,
            "knowledge": kg_result,
            "reasoning_steps": []
        }
    
    def generate(self, prompt: str) -> Dict[str, Any]:
        """生成多模态输出"""
        return self.multimodal.process_command(prompt)
    
    def process(self, input_text: str) -> Dict[str, Any]:
        """完整处理流程"""
        # 1. 理解
        understanding = self.understand(input_text)
        
        # 2. 推理
        reasoning = self.reason(input_text)
        
        # 3. 生成
        generation = self.generate(input_text)
        
        return {
            "understanding": understanding,
            "reasoning": reasoning,
            "generation": generation
        }


# 单元测试
if __name__ == "__main__":
    ai = UniCoreAI()
    
    # 测试理解
    result = ai.process("帮我搜索手机的最新价格")
    print(json.dumps(result, ensure_ascii=False, indent=2))
    
    # 测试推理
    ai.reasoning_engine.add_rule(r"价格", "返回价格信息")
    result = ai.reason("手机价格是多少")
    print(json.dumps(result, ensure_ascii=False, indent=2))

File: src/platform/web/js/ai_core.js
/**
 * UniCore AI层 - JavaScript实现
 * 意图理解、任务分解、知识推理
 */

class IntentClassifier {
    constructor() {
        this.INTENT_PATTERNS = {
            QUERY: ['什么', '怎么', '如何', '为什么', '多少', 'who', 'what', 'how', 'why'],
            EXECUTE: ['执行', '运行', '开始', '做', 'run', 'start', 'do'],
            CREATE: ['创建', '生成', '新建', '写', 'create', 'generate', 'new'],
            LEARN: ['学习', '教我', '告诉', '解释', 'learn', 'teach', 'explain'],
            CONTROL: ['控制', '开关', '调节', '设置', 'control', 'set', 'adjust'],
        };
        this.confidence_threshold = 0.6;
    }

    classify(text) {
        const scores = {};
        for (const intent in this.INTENT_PATTERNS) {
            scores[intent] = 0;
            for (const pattern of this.INTENT_PATTERNS[intent]) {
                if (text.toLowerCase().includes(pattern.toLowerCase())) {
                    scores[intent] += 1;
                }
            }
        }

        const total = Object.values(scores).reduce((a, b) => a + b, 0);
        if (total === 0) return { intent: 'UNKNOWN', confidence: 0 };

        for (const intent in scores) {
            scores[intent] /= total;
        }

        const bestIntent = Object.keys(scores).reduce((a, b) => scores[a] > scores[b] ? a : b);
        return { intent: bestIntent, confidence: scores[bestIntent] };
    }
}

class TaskDecomposer {
    constructor() {
        this.parallelKeywords = ['同时', '并行', '一起'];
    }

    decompose(task) {
        const subtasks = [];
        const parts = task.split(/[，,。. ]+/).filter(p => p.trim());

        let i = 0;
        for (const part of parts) {
            if (part.trim()) {
                const isParallel = this.parallelKeywords.some(k => task.includes(k));
                subtasks.push({
                    id: i,
                    description: part.trim(),
                    dependencies: isParallel ? [] : [i - 1],
                    parallel: isParallel && i > 0
                });
                i++;
            }
        }

        return subtasks;
    }
}

class EmbeddingModel {
    constructor(dim = 128) {
        this.dim = dim;
        this.vocab = {};
    }

    encode(text) {
        const words = Array.from(text);
        const vector = new Array(this.dim).fill(0);

        for (let i = 0; i < words.length; i++) {
            const char = words[i];
            let vec = this.vocab[char];
            if (!vec) {
                vec = Array.from({ length: this.dim }, (_, j) => 
                    Math.sin(char.charCodeAt(0) * (j + 1) / this.dim)
                );
                this.vocab[char] = vec;
            }
            for (let j = 0; j < this.dim; j++) {
                vector[j] += vec[j] / words.length;
            }
        }

        const norm = Math.sqrt(vector.reduce((a, b) => a + b * b, 0));
        if (norm > 0) {
            for (let i = 0; i < this.dim; i++) {
                vector[i] /= norm;
            }
        }

        return vector;
    }

    cosineSimilarity(a, b) {
        const dot = a.reduce((sum, val, i) => sum + val * b[i], 0);
        const normA = Math.sqrt(a.reduce((sum, val) => sum + val * val, 0));
        const normB = Math.sqrt(b.reduce((sum, val) => sum + val * val, 0));
        if (normA * normB === 0) return 0;
        return dot / (normA * normB);
    }
}

class ReasoningEngine {
    constructor() {
        this.rules = [];
        this.facts = {};
    }

    addRule(condition, action) {
        this.rules.push({ condition, action, pattern: new RegExp(condition) });
    }

    addFact(key, value) {
        this.facts[key] = value;
    }

    forwardChain(query) {
        for (const rule of this.rules) {
            if (rule.pattern.test(query)) {
                return rule.action;
            }
        }
        return null;
    }

    backwardChain(goal) {
        return this.rules.some(r => r.action === goal);
    }
}

class UniCoreAI {
    constructor() {
        this.intentClassifier = new IntentClassifier();
        this.taskDecomposer = new TaskDecomposer();
        this.embeddingModel = new EmbeddingModel();
        this.reasoningEngine = new ReasoningEngine();
        this.knowledgeGraph = new KnowledgeGraph();
        
        this.init();
    }

    init() {
        this.knowledgeGraph.addNode('system', { type: 'system', name: 'UniCore' });
        this.knowledgeGraph.addNode('platform', { type: 'platform', name: 'multi-platform' });
        this.knowledgeGraph.addEdge('system', 'platform', 'runs_on');
        
        this.reasoningEngine.addRule('价格', '返回价格信息');
        this.reasoningEngine.addRule('搜索', '执行搜索操作');
        this.reasoningEngine.addRule('创建', '执行创建操作');
    }

    understand(inputText) {
        const { intent, confidence } = this.intentClassifier.classify(inputText);
        const subtasks = this.taskDecomposer.decompose(inputText);

        return {
            text: inputText,
            intent,
            confidence,
            subtasks,
            embedding: this.embeddingModel.encode(inputText)
        };
    }

    reason(query) {
        const result = this.reasoningEngine.forwardChain(query);
        const kgResult = this.knowledgeGraph.query(query);

        return {
            result,
            knowledge: kgResult,
            reasoningSteps: []
        };
    }

    process(inputText) {
        const understanding = this.understand(inputText);
        const reasoning = this.reason(inputText);

        return {
            understanding,
            reasoning,
            response: this.generateResponse(understanding, reasoning)
        };
    }

    generateResponse(understanding, reasoning) {
        const responses = {
            QUERY: '我正在分析您的问题...',
            EXECUTE: '好的，正在执行您的命令...',
            CREATE: '正在为您创建...',
            LEARN: '让我来为您解释...',
            CONTROL: '正在控制设备...',
            UNKNOWN: '我不确定您的意图，请重新描述。'
        };

        return responses[understanding.intent] || responses.UNKNOWN;
    }
}

class KnowledgeGraph {
    constructor() {
        this.nodes = new Map();
        this.edges = [];
    }

    addNode(id, data) {
        this.nodes.set(id, data);
    }

    addEdge(fromId, toId, relation) {
        this.edges.push({ from: fromId, to: toId, relation });
    }

    query(nodeId) {
        const node = this.nodes.get(nodeId);
        const relations = this.edges.filter(e => e.from === nodeId).map(e => ({
            target: e.to,
            relation: e.relation,
            targetData: this.nodes.get(e.to)
        }));

        return { node, relations };
    }

    infer(start, end) {
        const visited = new Set();
        const queue = [[start, [start]]];

        while (queue.length > 0) {
            const [node, path] = queue.shift();
            if (node === end) return path;
            if (visited.has(node)) continue;
            visited.add(node);

            for (const edge of this.edges) {
                if (edge.from === node && !visited.has(edge.to)) {
                    queue.push([edge.to, [...path, edge.to]]);
                }
            }
        }

        return null;
    }
}

if (typeof module !== 'undefined' && module.exports) {
    module.exports = { UniCoreAI, IntentClassifier, TaskDecomposer, KnowledgeGraph };
}

if (typeof window !== 'undefined') {
    window.UniCoreAI = UniCoreAI;
    window.IntentClassifier = IntentClassifier;
    window.TaskDecomposer = TaskDecomposer;
    window.KnowledgeGraph = KnowledgeGraph;
}
